lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+      for (const auto &phTensorPair : phMap) {
+        auto *placeholder = phTensorPair.first;
+        const auto *tensor = phTensorPair.second;
+
+        if (symbolName == placeholder->getName()) {
+          executionState->insertIntoNodeCtx(node, placeholder, tensor->clone());
+        }
+      }
+    }
+  }
+}
+
+void ThreadPoolExecutor::executeDAGNode(RunIdentifierTy runId,
+                                        const DAGNode *node) {
+  // Get the execution state for the run.
+  auto executionStateIt = executionStates_.find(runId);
+
+  if (executionStateIt == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = executionStateIt->second;
+
+  // If execution has already failed due to another node, don't bother running
+  // this one.
+  if (executionState->getResultCode() == ResultCode::Failed) {
+    // Mark the node as no longer execution.
+    executionState->decrementInflightNodes();
+    return;
+  }
+
+  // Get the DeviceManager that can run the node.
+  auto deviceManagerIt = deviceManagers_.find(node->deviceID);
+
+  if (deviceManagerIt == deviceManagers_.end()) {
+    // Mark the node as no longer executing.
+    executionState->decrementInflightNodes();
+    executionState->setResultCode(ResultCode::Failed);
+    return;
+  }
+
+  std::shared_ptr<DeviceManager> deviceManager = deviceManagerIt->second;
+
+  // Get the Context containing all of the inputs for the node.
+  std::unique_ptr<Context> nodeCtx =
+      executionState->getUniqueNodeContextPtr(node);
+  // Run the node using the DeviceManager.
+  deviceManager->runFunction(
+      node->name, std::move(nodeCtx),
+      [this, runId, node](RunIdentifierTy id, ResultCode resultCode,
+                          std::unique_ptr<Context> resultCtx) {
+        // Immediately move the handling of the result onto threadPool_ to
+        // avoid doing work on the DeviceManager thread.
+        std::shared_ptr<Context> sharedResultCtx = std::move(resultCtx);
+        this->threadPool_.submit(
+            [this, runId, node, resultCode, sharedResultCtx]() {
+              this->handleDeviceManagerResult(runId, resultCode,
+                                              sharedResultCtx, node);
+            });
+      });
+}
+
+void ThreadPoolExecutor::propagateOutputPlaceholders(
+    RunIdentifierTy runId, std::shared_ptr<Context> ctx) {
+  // Get the execution state for the run.
+  auto it = executionStates_.find(runId);
+
+  if (it == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = it->second;
+
+  // Copy all of the Placeholders in ctx into the result Context for the run.
+  for (const auto &phTensorPair : ctx->pairs()) {
+    auto *placeholder = phTensorPair.first;
+    const auto *tensor = phTensorPair.second;
+
+    executionState->insertIntoResultCtx(placeholder, tensor->clone());
```

Comment:
Do we have to copy here? Can we move?

Reply:
Yep, unlike the other case, I think we can! Unless we want to allow tapping off intermediate results as outputs?